File: sdk/go/claim_test.go
package simplefin

import (
	"context"
	"encoding/base64"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestDecodeSetupToken(t *testing.T) {
	const claimURL = "https://bridge.example/simplefin/claim/abc123"
	token := base64.StdEncoding.EncodeToString([]byte(claimURL))

	t.Run("decodes", func(t *testing.T) {
		got, err := DecodeSetupToken(token)
		if err != nil {
			t.Fatalf("DecodeSetupToken: %v", err)
		}
		if got != claimURL {
			t.Errorf("got %q, want %q", got, claimURL)
		}
	})

	// Users paste these by hand out of a browser, so stray whitespace is the
	// normal case rather than the exceptional one.
	t.Run("tolerates surrounding whitespace", func(t *testing.T) {
		got, err := DecodeSetupToken("  \n\t" + token + "\n  ")
		if err != nil {
			t.Fatalf("DecodeSetupToken: %v", err)
		}
		if got != claimURL {
			t.Errorf("got %q, want %q", got, claimURL)
		}
	})

	t.Run("rejects bad input", func(t *testing.T) {
		bad := map[string]string{
			"empty":               "",
			"whitespace only":     "   ",
			"not base64":          "!!!not base64!!!",
			"base64 of nonsense":  base64.StdEncoding.EncodeToString([]byte("hello there")),
			"base64 of a non-URL": base64.StdEncoding.EncodeToString([]byte("ftp://example.org/claim")),
		}
		for name, input := range bad {
			t.Run(name, func(t *testing.T) {
				if _, err := DecodeSetupToken(input); err == nil {
					t.Errorf("DecodeSetupToken(%q) succeeded, want an error", input)
				}
			})
		}
	})
}

func TestClaim(t *testing.T) {
	const accessURL = "https://user:pass@bridge.example/simplefin"

	var gotMethod string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotMethod = r.Method
		// A trailing newline is common from servers that echo the URL.
		_, _ = w.Write([]byte(accessURL + "\n"))
	}))
	t.Cleanup(srv.Close)

	token := base64.StdEncoding.EncodeToString([]byte(srv.URL + "/claim/abc"))
	got, err := Claim(context.Background(), token, WithHTTPClient(srv.Client()))
	if err != nil {
		t.Fatalf("Claim: %v", err)
	}

	if gotMethod != http.MethodPost {
		t.Errorf("method = %q, want POST", gotMethod)
	}
	if got != accessURL {
		t.Errorf("access URL = %q, want %q", got, accessURL)
	}
}

func TestClaimErrors(t *testing.T) {
	t.Run("403 maps to ErrAuth", func(t *testing.T) {
		// Per the protocol checklist, a 403 here can mean the token was
		// already claimed by someone else and the user's data is compromised,
		// so it must be distinguishable from any other failure.
		srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusForbidden)
		}))
		t.Cleanup(srv.Close)

		token := base64.StdEncoding.EncodeToString([]byte(srv.URL + "/claim/abc"))
		_, err := Claim(context.Background(), token, WithHTTPClient(srv.Client()))
		if !errors.Is(err, ErrAuth) {
			t.Errorf("got %v, want ErrAuth", err)
		}
	})

	t.Run("non-URL body is rejected without echoing it", func(t *testing.T) {
		const secret = "oops-this-might-be-a-credential"
		srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			_, _ = w.Write([]byte(secret))
		}))
		t.Cleanup(srv.Close)

		token := base64.StdEncoding.EncodeToString([]byte(srv.URL + "/claim/abc"))
		_, err := Claim(context.Background(), token, WithHTTPClient(srv.Client()))
		if err == nil {
			t.Fatal("expected an error")
		}
		if !strings.Contains(err.Error(), "access URL") {
			t.Errorf("error = %q, want it to mention the access URL", err)
		}
		if strings.Contains(err.Error(), secret) {
			t.Errorf("error echoes the response body: %q", err)
		}
	})
}

File: sdk/node/test/claim.test.ts
import assert from "node:assert/strict"
import { createServer, type Server, type IncomingMessage, type ServerResponse } from "node:http"
import type { AddressInfo } from "node:net"
import { after, test } from "node:test"

import { claim, decodeSetupToken, isAuthError } from "../dist/index.js"

const servers: Server[] = []
after(() => {
  for (const s of servers) s.close()
})

async function startServer(
  handler: (req: IncomingMessage, res: ServerResponse) => void,
): Promise<string> {
  const server = createServer(handler)
  servers.push(server)
  await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", resolve))
  const { port } = server.address() as AddressInfo
  return `http://127.0.0.1:${port}`
}

const claimUrl = "https://bridge.example/simplefin/claim/abc123"
const token = btoa(claimUrl)

test("decodeSetupToken decodes a base64 URL", () => {
  assert.equal(decodeSetupToken(token), claimUrl)
})

// Users paste these by hand out of a browser, so stray whitespace is the
// normal case rather than the exceptional one.
test("decodeSetupToken tolerates surrounding whitespace", () => {
  assert.equal(decodeSetupToken(`  \n\t${token}\n  `), claimUrl)
})

test("decodeSetupToken rejects bad input", () => {
  const bad: Record<string, string> = {
    empty: "",
    "whitespace only": "   ",
    "not base64": "!!!not base64!!!",
    "base64 of nonsense": btoa("hello there"),
    "base64 of a non-URL": btoa("ftp://example.org/claim"),
  }
  for (const [name, input] of Object.entries(bad)) {
    assert.throws(() => decodeSetupToken(input), Error, name)
  }
})

test("claim POSTs and trims the returned access URL", async () => {
  const accessUrl = "https://user:pass@bridge.example/simplefin"
  let gotMethod = ""

  const origin = await startServer((req, res) => {
    gotMethod = req.method ?? ""
    // A trailing newline is common from servers that echo the URL.
    res.end(`${accessUrl}\n`)
  })

  const got = await claim(btoa(`${origin}/claim/abc`))

  assert.equal(gotMethod, "POST")
  assert.equal(got, accessUrl)
})

/*
 * Per the protocol checklist, a 403 here can mean the token was already
 * claimed by someone else and the user's data is compromised, so it must be
 * distinguishable from any other failure.
 */
test("a 403 during claim is an auth error", async () => {
  const origin = await startServer((_req, res) => {
    res.statusCode = 403
    res.end()
  })

  await assert.rejects(
    () => claim(btoa(`${origin}/claim/abc`)),
    (err: unknown) => {
      assert.ok(isAuthError(err), "expected an auth error")
      return true
    },
  )
})

test("a non-URL claim response is rejected without echoing it", async () => {
  const secret = "oops-this-might-be-a-credential"
  const origin = await startServer((_req, res) => res.end(secret))

  await assert.rejects(
    () => claim(btoa(`${origin}/claim/abc`)),
    (err: unknown) => {
      const msg = err instanceof Error ? err.message : String(err)
      assert.ok(msg.includes("access URL"), `error should mention the access URL: ${msg}`)
      assert.ok(!msg.includes(secret), `error echoes the response body: ${msg}`)
      return true
    },
  )
})
